cookbooks/tomography_based_plate_motions/plugins/reference_profile.cc

```diff
@@ -0,0 +1,457 @@
+      // point down / in and therefore gravity should be positive, leading to increasing
+      // adiabatic pressures and temperatures with depth. In some cases it will point up / out
+      // (e.g. for backward advection), in which case the pressures and temperatures should
+      // decrease with depth and therefore gravity has to be negative in the following equations.
+      const Tensor <1,dim> g = this->get_gravity_model().gravity_vector(this->get_geometry_model().representative_point(0));
+      const Point<dim> point_surf = this->get_geometry_model().representative_point(0);
+      const Point<dim> point_bot = this->get_geometry_model().representative_point(this->get_geometry_model().maximal_depth());
+      const int gravity_direction =  (g * (point_bot - point_surf) >= 0) ?
+                                     1 :
+                                     -1;
+
+      // We want to use different values for the densities in the uppermost vs the lower part of the mantle.
+      double uppermost_mantle_thickness = 0.0;
+      if (Plugins::plugin_type_matches<const MaterialModel::TomographyBasedPlateMotions<dim>>(this->get_material_model()))
+        {
+          const MaterialModel::TomographyBasedPlateMotions<dim> &material_model
+            = Plugins::get_plugin_as_type<const MaterialModel::TomographyBasedPlateMotions<dim>> (this->get_material_model());
+
+          uppermost_mantle_thickness = material_model.get_depth_of_base_of_uppermost_mantle();
+        }
+
+
+      // now integrate downward using the explicit Euler method for simplicity
+      //
+      // note: p'(z) = rho(p,T) * |g|
+      //       T'(z) = alpha |g| T / C_p
+      for (unsigned int i=0; i<n_points; ++i)
+        {
+          // use material properties calculated at i-1
+          // The first column in the input ascii file is the radius from the center of the Earth,
+          // therefore we convert the depths to the radial values in the get_data_component().
+          const double current_depth = double(i)/double(n_points-1)*this->get_geometry_model().maximal_depth();
+          const double outer_radius = this->get_geometry_model().representative_point(0.).norm();
+
+          double density = reference_profile.get_data_component(Point<1>(outer_radius - current_depth),density_index);
+
+          if (i==0)
+            {
+              if (!use_surface_condition_function)
+                {
+                  pressures[0] = this->get_surface_pressure();
+                  temperatures[0] = this->get_adiabatic_surface_temperature();
+                }
+              else
+                {
+                  pressures[0] = surface_condition_function.value(Point<1>(0.0),0);
+                  temperatures[0] = surface_condition_function.value(Point<1>(0.0),1);
+                }
+            }
+          else
+            {
+              // We only want to use the PREM densities in the part of the model that also
```

Comment:
Is this a copy-paste error? I thought you're not using PREM, or did I midunderstand?

Reply:
Not really, we are using PREM for defining our reference profile against the tomography. But, "reference densities" may be more appropriate.